tools/sim/bridge/metadrive.py

```diff
@@ -0,0 +1,130 @@
+      self.yuv = timer("yuv kernel", lambda: self.camerad.img_to_yuv(img))
+      timer("send yuv", lambda: self.camerad.cam_send_yuv_wide_road(self.yuv))
+      # timer("send yuv", lambda: self.camerad.cam_send_yuv_wide_road(self.yuv))
+
+  def get_velocity(self):
+    return None
+
+  def get_speed(self) -> float:
+    return self.speed
+
+  def get_steer_correction(self) -> float:
+    max_steer_angle = 75 / self.ticks_per_frame
+    return max_steer_angle * STEER_RATIO * -1
+  
+  def tick(self):
+    pass
+
+
+class MetaDriveBridge(SimulatorBridge):
+  TICKS_PER_FRAME = 10
+
+  def __init__(self, args):
+    if args.dual_camera:
+      warnings.warn("Dual camera not supported in MetaDrive simulator for performance reasons")
+      args.dual_camera = False
+    if args.ticks_per_frame:
+      self.TICKS_PER_FRAME = args.ticks_per_frame
+    super(MetaDriveBridge, self).__init__(args)
+
+  def spawn_objects(self):
+    # Lazily import as `metadrive-simulator`` is an optional install
+    import metadrive  # noqa: F401 pylint: disable=W0611, disable=import-error
+    import gym  # pylint: disable=import-error
+    from metadrive.constants import CamMask  # pylint: disable=import-error
+    from metadrive.component.vehicle_module.base_camera import BaseCamera  # pylint: disable=import-error
+    from metadrive.engine.engine_utils import engine_initialized  # pylint: disable=import-error
+    from metadrive.engine.core.image_buffer import ImageBuffer  # pylint: disable=import-error
+
+    env = gym.make('MetaDrive-10env-v0', config=dict(offscreen_render=True))
+
+    # config = dict(
+    #   # camera_dist=3.0,
+    #   # camera_height=1.0,
+    #   # use_render=True,
+    #   vehicle_config=dict(
+    #     # enable_reverse=True,
+    #     # image_source="rgb_camera",
+    #     # rgb_camera=(0,0)
+    #   ),
+    #   offscreen_render=True,
+    # )  
+    # from metadrive.utils.space import VehicleParameterSpace
+    # from metadrive.component.vehicle.vehicle_type import DefaultVehicle
+    # max_engine_force = VehicleParameterSpace.DEFAULT_VEHICLE["max_engine_force"]
+    # max_engine_force._replace(max=max_engine_force.max * 10)
+    # max_engine_force._replace(min=max_engine_force.min * 10)
+
+    env.reset()
+    class RGBCameraWide(BaseCamera):
+      # shape(dim_1, dim_2)
+      BUFFER_W = W  # dim 1
+      BUFFER_H = H  # dim 2
+      CAM_MASK = CamMask.RgbCam
+
+      def __init__(self):
+        assert engine_initialized(), "You should initialize engine before adding camera to vehicle"
+        self.BUFFER_W, self.BUFFER_H = W, H
+        super(RGBCameraWide, self).__init__()
+        cam = self.get_cam()
+        lens = self.get_lens()
+        cam.lookAt(0, 2.4, 1.3)
+        cam.setHpr(0, 0.8, 0)
+        lens.setFov(160)
+        lens.setAspectRatio(1.15)
+  
+    env.vehicle.add_image_sensor("rgb_wide", RGBCameraWide())
+
+    # Monkey patch `get_rgb_array` since it involves unnecessary copies
```

Comment:
looks like this has been merged upstream

Reply:
Let me check the perf

Reply:
Turns out upstream still does pretty bad things. Prefer to keep it as is.

Reply:
Seems fine actually.